### PR TITLE
Move claims to `BasicInterpreter`; misc refactorings

### PR DESCRIPTION
**Pull request by nishantjr**
This allows writing more powerful unit tests using the `BasicInterpreter`, by returning stubbed `ProofThunk`s.

### CODE REVIEW

**Review by IljaZakharov — APPROVED** [review]
Looks good! I left a couple of thoughts as comments.

File: generation/src/proof_generation/interpreter/basic_interpreter.py [hunk 1]
```diff
@@ -11,14 +11,18 @@
 if TYPE_CHECKING:
     from collections.abc import Mapping
 
-    from proof_generation.aml import Pattern
+    from ..aml import Pattern
```
[the remark this answers is not in the document]
Comment (nishantjr) [reply generation/src/proof_generation/interpreter/basic_interpreter.py:14]: I've been moving everything to relative imports as I go along. I've found that absolute imports are getting very cluttered and verbose. Sorry, I should have talked with the team before making this change.

File: generation/src/proof_generation/aml/utils.py [hunk 2]
```diff
@@ -70,7 +72,14 @@ def match_single(
         if pat_mu[0] != inst_mu[0]:
             return None
         return match_single(pat_mu[1], inst_mu[1], ret)
-    # TODO Consider adding cases for ESubst/SSubst
```
[the remark this answers is not in the document]
Comment (nishantjr) [reply generation/src/proof_generation/aml/utils.py:73]: Done!

File: generation/src/proof_generation/aml/syntax.py [hunk 3]
```diff
@@ -21,6 +22,7 @@ def evar_is_fresh(self, name: int) -> bool:
     def metavars(self) -> set[MetaVar]:
         raise NotImplementedError
 
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (nishantjr) [reply generation/src/proof_generation/aml/syntax.py:26]: Done!

File: generation/src/proof_generation/interpreter/basic_interpreter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nishantjr) [reply generation/src/proof_generation/interpreter/basic_interpreter.py:14]: It's also what pyk is using.

**Pull request merged**